File: plugins/social-media/skills/baoyu-post-to-wechat/scripts/publish-draft.py
#!/usr/bin/env python3
"""
WeChat draft publisher via CDP operate_appmsg API.

Key: must use type=77777 (new single-article draft), NOT type=77 (old material).
Old type=77 returns ret=0 but content goes to old material area, invisible in new draft box.

Usage:
  python publish-draft.py --title "标题" --file body.html [--digest "摘要"] [--del OLD_ID]

Requirements: pip install websockets
"""

import argparse, asyncio, json, os, sys, urllib.request

CDP_PORT = 9222


async def get_ws_url():
    """Get WeChat tab's WebSocket URL from Chrome DevTools."""
    try:
        tabs = json.loads(urllib.request.urlopen(f"http://127.0.0.1:{CDP_PORT}/json").read())
    except Exception as e:
        print(f"❌ Cannot connect to Chrome on port {CDP_PORT}: {e}", file=sys.stderr)
        sys.exit(1)

    wx_tab = None
    for t in tabs:
        if "mp.weixin.qq.com" in t.get("url", ""):
            wx_tab = t
            break

    if not wx_tab:
        print("❌ No WeChat tab found. Navigate to https://mp.weixin.qq.com first.", file=sys.stderr)
        sys.exit(1)

    return wx_tab["webSocketDebuggerUrl"]


async def cdp_eval(ws, js_expr, msg_id=1):
    """Execute JS in page context via CDP Runtime.evaluate, return result value."""
    await ws.send(json.dumps({
        "id": msg_id, "method": "Runtime.evaluate",
        "params": {"expression": js_expr, "returnByValue": True, "timeout": 30000}
    }))
    resp = json.loads(await ws.recv())
    return resp.get("result", {}).get("result", {}).get("value", "")


async def cdp_eval_raw(ws, js_expr, msg_id=1):
    """Execute JS in page context, return raw response dict."""
    await ws.send(json.dumps({
        "id": msg_id, "method": "Runtime.evaluate",
        "params": {"expression": js_expr, "returnByValue": True, "timeout": 30000}
    }))
    return json.loads(await ws.recv())


async def get_token(ws):
    """Get WeChat session token with multiple fallback sources."""
    js = """(function() {
        if (window.__token) return window.__token;
        var m = document.cookie.match(/token=([^;]+)/);
        if (m) return m[1];
        m = location.href.match(/token=([^&]+)/);
        if (m) return m[1];
        return '';
    })()"""
    token = await cdp_eval(ws, js, msg_id=1)
    if not token:
        print("❌ No token found. Need to scan QR code to login.", file=sys.stderr)
        sys.exit(1)
    return token


async def delete_draft(ws, token, draft_id):
    """Delete an existing draft by appMsgId."""
    js = f"""(function() {{
        var x = new XMLHttpRequest();
        x.open('POST', 'https://mp.weixin.qq.com/cgi-bin/operate_appmsg?t=ajax-response&sub=del&type=77&token={token}&lang=zh_CN', false);
        x.setRequestHeader('Content-Type', 'application/x-www-form-urlencoded; charset=UTF-8');
        x.send('token={token}&lang=zh_CN&f=json&ajax=1&AppMsgId={draft_id}&count=1&data_seq=0');
        return x.responseText;
    }})()"""
    await cdp_eval(ws, js, msg_id=10)
    print(f"Deleted draft {draft_id}")


async def create_draft(ws, token, title, body_html, digest=""):
    """Create a new draft via operate_appmsg API with type=77777 (new single-article).

    Critical: use FormData (multipart), NOT URL-encoded string.
    Must store body in window first to avoid JS string escaping issues.
    """
    # Step 1: Store body_html and title in window context
    store_js = f"window.__draft_body__ = {json.dumps(body_html)}; window.__draft_title__ = {json.dumps(title)}; window.__draft_digest__ = {json.dumps(digest)}; 'stored'"
    result = await cdp_eval(ws, store_js, msg_id=20)
    if result != "stored":
        print(f"❌ Failed to store content in page context: {result}", file=sys.stderr)
        return None

    # Step 2: Create draft via synchronous XHR + FormData + type=77777
    create_js = f"""(function() {{
        var x = new XMLHttpRequest();
        x.open('POST', 'https://mp.weixin.qq.com/cgi-bin/operate_appmsg?sub=create&type=77777&token={token}', false);
        var fd = new FormData();
        fd.append('token', '{token}');
        fd.append('lang', 'zh_CN');
        fd.append('f', 'json');
        fd.append('ajax', '1');
        fd.append('AppMsgId', '');
        fd.append('count', '1');
        fd.append('data_seq', '0');
        fd.append('title0', window.__draft_title__);
        fd.append('content0', window.__draft_body__);
        fd.append('digest0', window.__draft_digest__);
        fd.append('author0', '');
        fd.append('content_source_url0', '');
        fd.append('need_open_comment0', '0');
        fd.append('show_cover_pic0', '1');
        fd.append('copyright_type0', '0');
        fd.append('can_reward0', '0');
        fd.append('mediaapi_publish_status0', '0');
        fd.append('fee_type0', '');
        fd.append('pay_fee0', '0');
        fd.append('pay_album_info0', '');
        fd.append('is_set_sync_to_finder0', '0');
        x.send(fd);
        return x.responseText;
    }})()"""

    result_str = await cdp_eval(ws, create_js, msg_id=21)
    return result_str


async def main():
    parser = argparse.ArgumentParser(description="Publish WeChat draft via CDP")
    parser.add_argument("--title", required=True, help="Article title (≤64 chars)")
    parser.add_argument("--file", required=True, help="Path to body HTML file")
    parser.add_argument("--digest", default="", help="Article digest (≤120 chars)")
    parser.add_argument("--del", dest="delete_id", help="Draft appMsgId to delete before creating")
    args = parser.parse_args()

    # Read body HTML
    if not os.path.exists(args.file):
        print(f"❌ File not found: {args.file}", file=sys.stderr)
        sys.exit(1)

    with open(args.file, "r", encoding="utf-8") as f:
        body_html = f.read().strip()

    print(f"body: {len(body_html)}")

    # Install websockets if needed
    try:
        import websockets
    except ImportError:
        import subprocess
        subprocess.check_call([sys.executable, "-m", "pip", "install", "websockets", "-q"])
        import websockets

    # Connect to Chrome
    ws_url = await get_ws_url()

    async with websockets.connect(ws_url, max_size=50 * 1024 * 1024) as ws:
        # Get token
        token = await get_token(ws)
        print(f"token: {token}")

        # Delete old draft if requested
        if args.delete_id:
            print(f"Deleting draft {args.delete_id}...")
            await delete_draft(ws, token, args.delete_id)

        # Create draft
        print("Creating draft...")
        result_str = await create_draft(ws, token, args.title, body_html, args.digest)

        if not result_str:
            print("❌ Empty response from API", file=sys.stderr)
            sys.exit(1)

        try:
            data = json.loads(result_str)
        except json.JSONDecodeError as e:
            print(f"❌ Cannot parse API response: {e}", file=sys.stderr)
            print(f"   Raw: {result_str[:300]}", file=sys.stderr)
            sys.exit(1)

        ret = data.get("ret")
        base_ret = data.get("base_resp", {}).get("ret")
        app_msg_id = data.get("appMsgId") or data.get("appmsgid")
        err_msg = data.get("base_resp", {}).get("err_msg", "")

        if ret == 0 or ret == "0" or base_ret == 0 or base_ret == "0":
            content_len = 0
            filtered = data.get("filter_content_html", [])
            if filtered and filtered[0].get("content"):
                content_len = len(filtered[0]["content"])
            print(f"✅ {app_msg_id} content: {content_len}")
        else:
            print(f"❌ ret={ret} base_ret={base_ret} msg={err_msg}")
            sys.exit(1)


if __name__ == "__main__":
    asyncio.run(main())

File: plugins/social-media/skills/baoyu-post-to-wechat/templates/publish-draft.mjs
#!/usr/bin/env bun
// WeChat draft publisher via CDP operate_appmsg API
// Usage: bun publish-draft.mjs --title "标题" --file body.html [--digest "摘要"] [--del OLD_ID]
//
// Usage:
//   bun /tmp/publish-draft.mjs --title "My Title" --file /path/to/body.html
//   bun /tmp/publish-draft.mjs --title "My Title" --file /path/to/body.html --del 100000123

const args = process.argv.slice(2);
function getArg(name) {
  const idx = args.indexOf('--' + name);
  return idx >= 0 && idx + 1 < args.length ? args[idx + 1] : null;
}

const BODY_FILE = getArg('file');
const TITLE = getArg('title');
const DIGEST = getArg('digest') || '';
const DEL_ID = getArg('del'); // optional: draft ID to delete before creating

if (!BODY_FILE || !TITLE) {
  console.error('Usage: bun publish-draft.mjs --title "标题" --file body.html [--digest "摘要"] [--del OLD_ID]');
  process.exit(1);
}

class BunCdp {
  ws=null; pending=new Map(); nextId=1;
  connect(wsUrl) {
    const self=this;
    return new Promise(function(res,rej) {
      const t=setTimeout(function(){rej(new Error('connect timeout'))},15000);
      self.ws=new WebSocket(wsUrl);
      self.ws.addEventListener('open',function(){clearTimeout(t);res()});
      self.ws.addEventListener('message',function(evt){
        try{var msg=JSON.parse(evt.data);
          if(msg.id){var p=self.pending.get(msg.id);if(p){self.pending.delete(msg.id);clearTimeout(p.t);p.res(msg.result);}}
        }catch{}
      });
      self.ws.addEventListener('error',function(){clearTimeout(t);rej(new Error('WS error'))});
    });
  }
  send(m,params,opts) {
    const self=this; params=params||{};
    var sid = opts&&opts.sessionId;
    return new Promise(function(res,rej) {
      if(!self.ws){rej(new Error('not connected'));return;}
      const id=self.nextId++;
      const t=setTimeout(function(){self.pending.delete(id);rej(new Error(m+' timeout'))},120000);
      self.pending.set(id,{res:res,rej:rej,t:t});
      self.ws.send(JSON.stringify(sid?{id:id,method:m,params:params,sessionId:sid}:{id:id,method:m,params:params}));
    });
  }
}

async function ev(cdp,expr,sid) {
  var r = await cdp.send("Runtime.evaluate",{expression:expr,returnByValue:true},{sessionId:sid});
  return r.result?r.result.value:undefined;
}

async function main() {
  const {readFileSync} = await import("fs");

  // 1. Connect to Chrome
  var ver = await (await fetch("http://127.0.0.1:9222/json/version")).json();
  var cdp = new BunCdp();
  await cdp.connect(ver.webSocketDebuggerUrl);

  // 2. Find WeChat tab
  var tabs = await (await fetch("http://127.0.0.1:9222/json")).json();
  var wxTab = tabs.find(function(t){return t.url.includes("mp.weixin.qq.com")});
  if (!wxTab) { console.error("No WeChat tab found"); process.exit(1); }
  var sid = (await cdp.send("Target.attachToTarget",{targetId:wxTab.id,flatten:true})).sessionId;
  // Get real token from page context (Chrome strips token=*** from debugger URL)
  var token = await ev(cdp, "(window.__token || document.cookie.match(/token=([^;]+)/)?.[1] || location.href.match(/token=([^&]+)/)?.[1])", sid);
  if (!token) { console.error("Could not get token from page"); process.exit(1); }
  console.log("token obtained from page context");

  // 3. Optionally delete old draft
  if (DEL_ID) {
    console.log("[1] Del " + DEL_ID + "...");
    await ev(cdp, "(function(){var x=new XMLHttpRequest();x.open('POST','https://mp.weixin.qq.com/cgi-bin/operate_appmsg?t=ajax-response&sub=del&type=77&token="+token+"&lang=zh_CN',false);x.setRequestHeader('Content-Type','application/x-www-form-urlencoded; charset=UTF-8');x.send('token="+token+"&lang=zh_CN&f=json&ajax=1&AppMsgId="+DEL_ID+"&count=1&data_seq=0')})()", sid).catch(function(){});
  }

  // 4. Read body HTML
  var body = readFileSync(BODY_FILE,"utf-8").trim();
  console.log("body:",body.length);

  // 5. Create draft
  console.log("Creating draft...");
  var r = await ev(cdp, "(function(){var x=new XMLHttpRequest();x.open('POST','https://mp.weixin.qq.com/cgi-bin/operate_appmsg?t=ajax-response&sub=create&type=77&token="+token+"&lang=zh_CN',false);x.setRequestHeader('Content-Type','application/x-www-form-urlencoded; charset=UTF-8');var p={token:'"+token+"',lang:'zh_CN',f:'json',ajax:'1',random:Math.random().toString(),AppMsgId:'',count:'1',data_seq:'0',operate_from:'Chrome',isMark:'0'};p['title0']="+JSON.stringify(TITLE)+";p['content0']="+JSON.stringify(body)+";p['author0']='';p['fileid0']='';p['digest0']="+JSON.stringify(DIGEST)+";p['sourceurl0']='';p['need_open_comment0']='0';p['show_cover_pic0']='1';p['copyright_type0']='0';p['can_reward0']='0';p['mediaapi_publish_status0']='0';p['fee_type0']='';p['pay_fee0']='0';p['pay_album_info0']='';p['is_set_sync_to_finder0']='0';var fd=Object.keys(p).map(function(k){return encodeURIComponent(k)+'='+encodeURIComponent(p[k])}).join('&');x.send(fd);return x.responseText})()", sid);

  var resp = JSON.parse(r);
  console.log("ret:",resp.ret,"id:",resp.appMsgId);
  if(resp.ret==0||resp.ret=="0"){
    var ct=resp.filter_content_html&&resp.filter_content_html[0]&&resp.filter_content_html[0].content;
    console.log("✅",resp.appMsgId,"content:",ct?ct.length:0);
  } else {
    console.log("❌",JSON.stringify(resp).substring(0,200));
  }
}
main().catch(function(e){console.error(e);process.exit(1)});
